src/AbstractActiveRecord.php

```diff
@@ -1096,15 +1070,17 @@ protected function deleteInternal(): int
          * the database and thus the method will return 0
          */
         $condition = $this->getOldPrimaryKey(true);
-        $lock = $this->optimisticLock();
 
-        if ($lock !== null) {
+        if ($this instanceof OptimisticLockInterface) {
+            $lock = $this->optimisticLock();
             $condition[$lock] = $this->get($lock);
 
             $result = $this->deleteAll($condition);
 
             if ($result === 0) {
-                throw new StaleObjectException('The object being deleted is outdated.');
+                throw new OptimisticLockException(
+                    'The object being deleted is outdated. The record was changed by another process.'
```

Comment:
```suggestion
                    'The object being deleted is outdated.'
```

I think this message is enough and details could be confusing.